### PR TITLE
feature: added support for batch transform with model monitoring

### PR DESCRIPTION
**Pull request by keshav-chandak**
*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/model_monitor/model_monitoring.py [hunk 1]
```diff
@@ -3382,6 +3447,109 @@ def _to_request_dict(self):
...
+        inference_attribute: str = None,
+        probability_attribute: str = None,
+        probability_threshold_attribute: str = None,
+    ):
+        """Initialize a `BatchTransformInput` instance.
+
+        Args:
+            data_captured_destination_s3_uri (str): Location to the batch transform data
+                file which needs to be analysed.
+            destination (str): The output path for monitored data output.
```
Comment (jerrypeng7773) [inline src/sagemaker/model_monitor/model_monitoring.py:3492]: destination (str): The destination of the input.

**Review by jerrypeng7773 — CHANGES_REQUESTED** [review]
nit for the two parameter docstring. Other than that, LGTM.

**Review by jerrypeng7773 — APPROVED** [review]

File: src/sagemaker/inputs.py [hunk 2]
```diff
@@ -232,3 +233,43 @@ def __init__(
...
+        generate_inference_id: bool = None,
+    ):
+        """Create new BatchDataCaptureConfig
+
+        Args:
+            destination_s3_uri (str): S3 Location to store the captured data
+            kms_key_id (str): The KMS key to use when writing to S3.
+                KmsKeyId can be an ID of a KMS key, ARN of a KMS key, alias of a KMS key,
+                or alias of a KMS key. The KmsKeyId is applied to all outputs.
+            generate_inference_id (bool): Flag to generate an inference id
```
Comment (mufaddal-rohawala) [inline src/sagemaker/inputs.py:258]: add `(dafault: None)`

File: src/sagemaker/model_monitor/clarify_model_monitoring.py [hunk 3]
```diff
@@ -500,14 +524,15 @@ def suggest_baseline(
     # noinspection PyMethodOverriding
     def create_monitoring_schedule(
         self,
-        endpoint_input,
         ground_truth_input,
+        endpoint_input=None,
```
Comment (mufaddal-rohawala) [inline src/sagemaker/model_monitor/clarify_model_monitoring.py:528]: changing the order of arguments might break for Cx's existing workflows.

File: src/sagemaker/model_monitor/model_monitoring.py [hunk 4]
```diff
@@ -1498,7 +1518,7 @@ def suggest_baseline(
 
     def create_monitoring_schedule(
         self,
-        endpoint_input,
+        endpoint_input=None,
```
Comment (mufaddal-rohawala) [inline src/sagemaker/model_monitor/model_monitoring.py:1521]: add this to docstring

File: src/sagemaker/model_monitor/model_monitoring.py [hunk 5]
```diff
@@ -1540,6 +1561,8 @@ def create_monitoring_schedule(
                 expressions. Default: Daily.
             enable_cloudwatch_metrics (bool): Whether to publish cloudwatch metrics as part of
                 the baselining or monitoring jobs.
+            batch_transform_input (sagemaker.model_monitor.BatchTransformInput): Inputs to
```
Comment (mufaddal-rohawala) [inline src/sagemaker/model_monitor/model_monitoring.py:1564]: default

File: src/sagemaker/model_monitor/model_monitoring.py [hunk 6]
```diff
@@ -1844,6 +1878,8 @@ def _update_data_quality_monitoring_schedule(
             network_config (sagemaker.network.NetworkConfig): A NetworkConfig
                 object that configures network isolation, encryption of
                 inter-container traffic, security group IDs, and subnets.
+            batch_transform_input (sagemaker.model_monitor.BatchTransformInput): Inputs to
```
Comment (mufaddal-rohawala) [inline src/sagemaker/model_monitor/model_monitoring.py:1881]: default

File: src/sagemaker/model_monitor/model_monitoring.py [hunk 7]
```diff
@@ -3382,6 +3447,109 @@ def _to_request_dict(self):
...
+    It also provides a method to turn those parameters into a dictionary.
+
+    """
+
+    start_time_offset: str = attr.ib()
+    end_time_offset: str = attr.ib()
+    features_attribute: str = attr.ib()
+    inference_attribute: str = attr.ib()
+    probability_attribute: Union[str, int] = attr.ib()
+    probability_threshold_attribute: float = attr.ib()
```
Comment (mufaddal-rohawala) [inline src/sagemaker/model_monitor/model_monitoring.py:3464]: Need to add docstrings on these

File: src/sagemaker/model_monitor/model_monitoring.py [hunk 8]
```diff
@@ -3382,6 +3447,109 @@ def _to_request_dict(self):
...
+            start_time_offset (str): Monitoring start time offset, e.g. "-PT1H"
+            end_time_offset (str): Monitoring end time offset, e.g. "-PT0H".
+            features_attribute (str): JSONpath to locate features in JSONlines dataset.
+                Only used for ModelBiasMonitor and ModelExplainabilityMonitor
+            inference_attribute (str): Index or JSONpath to locate predicted label(s).
+                Only used for ModelQualityMonitor, ModelBiasMonitor, and ModelExplainabilityMonitor
+            probability_attribute (str): Index or JSONpath to locate probabilities.
+                Only used for ModelQualityMonitor, ModelBiasMonitor and ModelExplainabilityMonitor
+            probability_threshold_attribute (float): threshold to convert probabilities to binaries
+                Only used for ModelQualityMonitor, ModelBiasMonitor and ModelExplainabilityMonitor
```
Comment (mufaddal-rohawala) [inline src/sagemaker/model_monitor/model_monitoring.py:3505]: add defaults

File: src/sagemaker/session.py [hunk 9]
```diff
@@ -2577,6 +2584,8 @@ def transform(
             model_client_config (dict): A dictionary describing the model configuration for the
                 job. Dictionary contains two optional keys,
                 'InvocationsTimeoutInSeconds', and 'InvocationsMaxRetries'.
+            batch_data_capture_config (BatchDataCaptureConfig): Configuration object which
```
Comment (mufaddal-rohawala) [inline src/sagemaker/session.py:2587]: defaults

**Review by mufaddal-rohawala — CHANGES_REQUESTED** [review]

File: src/sagemaker/model_monitor/clarify_model_monitoring.py (hunk 3 above)
Comment (jerrypeng7773) [reply src/sagemaker/model_monitor/clarify_model_monitoring.py:528]: yea, but endpoint_input became optional, and non-none `ground_truth_input` has to come first.

File: src/sagemaker/inputs.py (hunk 2 above)
Comment (jerrypeng7773) [reply src/sagemaker/inputs.py:258]: done

File: src/sagemaker/model_monitor/model_monitoring.py (hunk 4 above)
Comment (jerrypeng7773) [reply src/sagemaker/model_monitor/model_monitoring.py:1521]: it exists in the docstring.

File: src/sagemaker/model_monitor/model_monitoring.py (hunk 7 above)
Comment (jerrypeng7773) [reply src/sagemaker/model_monitor/model_monitoring.py:3475]: done

File: src/sagemaker/model_monitor/model_monitoring.py (hunk 8 above)
Comment (jerrypeng7773) [reply src/sagemaker/model_monitor/model_monitoring.py]: done

File: src/sagemaker/session.py (hunk 9 above)
Comment (jerrypeng7773) [reply src/sagemaker/session.py:2588]: done

File: src/sagemaker/inputs.py (hunk 2 above)
Comment (jerrypeng7773) [reply src/sagemaker/inputs.py]: done

**Review by vikas0203 — COMMENTED** [review]
/bot run all

File: src/sagemaker/model_monitor/clarify_model_monitoring.py (hunk 3 above)
Comment (jerrypeng7773) [reply src/sagemaker/model_monitor/clarify_model_monitoring.py:527]: talked offline with @mufaddal-rohawala, we default `ground_truth_input` to `None` in the signature, however, we raise value error if it's None within the function.

File: src/sagemaker/model_monitor/model_monitoring.py [hunk 10]
```diff
@@ -217,8 +220,9 @@ def run_baseline(
 
     def create_monitoring_schedule(
```
Comment (mufaddal-rohawala) [inline src/sagemaker/model_monitor/model_monitoring.py:221]: Can we maintain:
```
def create_monitoring_schedule(
        self,
        endpoint_input=None,
	output=None,
        statistics=None,
        constraints=None,
        monitor_schedule_name=None,
        schedule_cron_expression=None,
        batch_transform_input=None,                              # new argument in the end and maintain sequence
    ):
```

**Review by mufaddal-rohawala — CHANGES_REQUESTED** [review]

**Review by claytonparnell — COMMENTED** [review]
/bot run slow-tests

**Pull request merged**